objc2/src/foundation/mutable_set.rs

```diff
@@ -0,0 +1,303 @@
+use alloc::vec::Vec;
+use core::fmt;
+use core::marker::PhantomData;
+
+use super::set::with_objects;
+use super::{NSCopying, NSFastEnumeration, NSFastEnumerator, NSMutableCopying, NSObject, NSSet};
+use crate::rc::{DefaultId, Id, Owned, Ownership, Shared, SliceId};
+use crate::{ClassType, Message, __inner_extern_class, extern_methods, msg_send, msg_send_id};
+
+__inner_extern_class!(
+    /// A growable unordered collection of unique objects.
+    ///
+    /// See the documentation for [`NSSet`] and/or [Apple's
+    /// documentation][apple-doc] for more information.
+    ///
+    /// [apple-doc]: https://developer.apple.com/documentation/foundation/nsmutableset?language=objc
+    #[derive(PartialEq, Eq, Hash)]
+    pub struct NSMutableSet<T: Message> {
+        item: PhantomData<Id<T, Shared>>,
+    }
+
+    unsafe impl<T: Message> ClassType for NSMutableSet<T> {
+        #[inherits(NSObject)]
+        type Super = NSSet<T>;
+    }
+);
+
+// SAFETY: Same as NSSet<T>
+unsafe impl<T: Message + Sync + Send> Sync for NSMutableSet<T> {}
+unsafe impl<T: Message + Sync + Send> Send for NSMutableSet<T> {}
+
+extern_methods!(
+    unsafe impl<T: Message> NSMutableSet<T> {
+        /// Creates an empty `NSMutableSet`.
+        ///
+        /// # Examples
+        ///
+        /// ```
+        /// use objc2::foundation::{NSMutableSet, NSString};
+        /// # #[cfg(feature = "gnustep-1-7")]
+        /// # unsafe { objc2::__gnustep_hack::get_class_to_force_linkage() };
+        ///
+        /// let set = NSMutableSet::<NSString>::new();
+        /// ```
+        pub fn new() -> Id<Self, Owned> {
+            // SAFETY:
+            // Same as `NSSet::new`, except mutable sets are always unique.
+            unsafe { msg_send_id![Self::class(), new] }
+        }
+
+        /// Creates an `NSMutableSet` from a vector.
+        ///
+        /// # Examples
+        ///
+        /// ```
+        /// use objc2::foundation::{NSMutableSet, NSString};
+        /// # #[cfg(feature = "gnustep-1-7")]
+        /// # unsafe { objc2::__gnustep_hack::get_class_to_force_linkage() };
+        ///
+        /// let strs = ["one", "two", "three"].map(NSString::from_str).to_vec();
+        /// let set = NSMutableSet::from_vec(strs);
+        /// ```
+        pub fn from_vec<O: Ownership>(vec: Vec<Id<T, O>>) -> Id<Self, Owned> {
+            // SAFETY:
+            // We always return `Id<NSMutableSet<T, Shared>, Owned>` because
+            // mutable sets are always unique and allow adding/removing elements
+            // but prevent modifying elements.
```

Comment:
I'm unsure this is the right tradeoff with mutable sets (also the main reason why `NSArray` and `NSMutableArray` have an ownership parameter; I'm unsure what the right tradeoff is!)

While it makes sense for e.g. `NSString` (which is always `Shared`), I can easily imagine counter-examples where users want to store their elements in a set, and mutably retrieve them afterwards.

Reply:
My thinking was that if users were able to modify the elements (without the set knowing about it) it would invalidate the set. I think that's why Rust's `HashSet`, unlike `Vec`, only has `get` and not `get_mut`.

According to [Apple's NSSet documentation](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Collections/Articles/Sets.html#//apple_ref/doc/uid/20000136-SW1):

> If mutable objects are stored in a set, either the hash method of the objects shouldn’t depend on the internal state of the mutable objects or the mutable objects shouldn’t be modified while they’re in the set.

It's possible for us to prevent users from modifying objects in the set but I don't know how we'd make sure the hash method of mutable objects doesn't depend on its internal state. Maybe it would be enough to document these requirements in `NSSet::get_mut` because I do agree it could be useful.

Reply:
Oh, right... Duh, forgot how sets work for a moment!

Though an `Id<NSSet, Owned>` would have the property that you could get mutable objects _back_:
```rust
let objs: Vec<Id<NSMutableString, Owned>> = ...; // A list of mutable strings
let set: Id<NSSet<NSMutableString, Owned>, Owned> = NSSet::from_vec(objs); // Deduplicate strings
let objs: Vec<Id<NSMutableString, Owned>> = set.into_vec(); // Get mutable/owned strings back
```

Reply:
That's a good point about getting mutable objects back through `NSSet::into_vec`. I'll give `NSSet` an ownership parameter to make it more like `NSArray`.